pkg/adapterManager/manager2.go

```diff
@@ -0,0 +1,439 @@
+		func(_ *HandlerState, inst *cpb.Instance) bool {
+			return inst.Name == qma.Quota
+		})
+	if err != nil {
+		return nil, err
+	}
+
+	ra := make([]*runArg, 0, 10)
+
+	// TODO reduce context deadline before sending it down to adapters.
+	for _, call := range calls {
+		for _, inst := range call.instances {
+			ra = append(ra, &runArg{call, func(ctx context.Context) *result2 {
+				resp, err := call.processor.ProcessQuota(ctx, inst.Name,
+					inst.Params.(proto.Message), requestBag, m.mapper, call.handlerState.Handler,
+					adapter.QuotaRequestArgs{
+						DeduplicationID: qma.DeduplicationID,
+						QuotaAmount:     qma.Amount,
+						BestEffort:      qma.BestEffort,
+					})
+				return &result2{err, &resp, call}
+			}})
+		}
+	}
+
+	qres, err := m.run(ctx, ra)
+	var res *adapter.QuotaResult2
+	if qres != nil {
+		res = qres.(*adapter.QuotaResult2)
+	}
+	return res, err
+}
+
+// Preprocess runs the first phase of adapter processing before any other adapters are run.
+// Attribute producing adapters are run in this phase.
+func (m *Manager2) Preprocess(ctx context.Context, requestBag attribute.Bag, responseBag *attribute.MutableBag) error {
+	// FIXME
+	return errors.New("not implemented")
+}
+
+// combineResults2 combines results
+func combineResults2(results []*result2) (adapter.Result, error) {
+	var res adapter.Result
+	var err *multierror.Error
+	var buf *bytes.Buffer
+	code := rpc.OK
+
+	for _, rs := range results {
+		if rs.err != nil {
+			err = multierror.Append(err, rs.err)
+		}
+
+		if rs.res == nil { // When there is no return value like ProcessReport().
+			continue
+		}
+		if res == nil {
+			res = rs.res
+		} else {
+			res.Combine(rs.res)
+		}
+		st := rs.res.GetStatus()
+		if !status.IsOK(st) {
+			if buf == nil {
+				buf = pool.GetBuffer()
+				// the first failure result's code becomes the result code for the output
+				code = rpc.Code(st.Code)
+			} else {
+				buf.WriteString(", ")
+			}
+
+			buf.WriteString(rs.callinfo.handlerName + ":" + st.Message)
+		}
+	}
+
+	if buf != nil {
+		res.SetStatus(status.WithMessage(code, buf.String()))
+		pool.PutBuffer(buf)
+	}
+	return res, err.ErrorOrNil()
+}
+
+// doFunc is the abstraction used by runAsync to do work.
+type doFunc func(context.Context) *result2
+
+// result2 encapsulates commonalities between all adapter returns.
+type result2 struct {
+	// all results return an error
+	err error
+	// CheckResult or QuotaResult
+	res adapter.Result
+	// callinfo that resulted in "res"
+	callinfo *dispatchInfo
+}
+
+// runArg encapsulates callinfo with the doFunc that acts on it.
+type runArg struct {
+	callinfo *dispatchInfo
+	do       doFunc
+}
+
+// run runArgs using runAsync and return results.
+func (m *Manager2) run(ctx context.Context, runArgs []*runArg) (adapter.Result, error) {
+	nresults := len(runArgs)
+	resultsChan := make(chan *result2, nresults)
+	results := make([]*result2, nresults)
+
+	for _, ra := range runArgs {
+		m.runAsync(ctx, ra.callinfo, resultsChan, ra.do)
```

Comment:
I have refactored so that Check/Report/Quota are smaller and dispatch does most of the work.